[Dynamo] Fix nested function resume execution

yanboliang opened the pull request:
Fixes #99665

Let me explain the root cause using the unit test I added:
* This bug is triggered when:
  * ```wrapped``` is a nested function.
  * ```wrapped``` is in another module which is different from the main function ```fn```.
  * There is a graph break inside of ```wrapped```.
* The root cause is when resuming nested function, actually we are using the outermost function(```fn``` in my example)'s global variables, but ```wrapped``` calls ```inner_func``` which is not part of ```fn```'s globals, so we have to set correct globals when nested function resume execution.


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

yanboliang commented on test/dynamo/utils.py:
```diff
@@ -0,0 +1,17 @@
+# Owner(s): ["module: dynamo"]
```
Will refactor out other cross imports into this file as a follow-up PR.

jansel commented on torch/_dynamo/variables/functions.py:
```diff
@@ -492,6 +492,19 @@ def reconstruct(self, codegen):
             codegen(self.fn_name)
         codegen.extend_output([create_instruction("MAKE_FUNCTION", arg=flags)])
 
+        # When graph breaks happen, the global variables of resumed function
+        # come from the outermost function which may not have globals needed
+        # by this nested function, we have to add these missing globals.
+        root_globals = codegen.tx.output.root_globals
+        for key in self.code.value.co_names:
+            if key not in self.f_globals:
+                continue
+            if key in root_globals:
+                if not (key.startswith("__") and key.endswith("__")):
+                    assert root_globals[key] == self.f_globals[key]
+            else:
+                root_globals[key] = self.f_globals[key]
+
```
This is modifying the user's global scope, copying every object from one user module to another user module.  This is not correct and could break user code.

The root problem is that f_globals is pointing to the wrong module.  Which I believe is coming from the usage of the `MAKE_FUNCTION` bytecode above.

Instead of using `MAKE_FUNCTION`, I think we can call:
```
types.FunctionType(code, globals, name, argdefs, closure)
```
directly.  The constructor of `FunctionType` allows passing a globals pointer different from the current module.

anijain2305 commented:
Very cool!

jansel commented on torch/_dynamo/variables/functions.py:
```diff
@@ -460,17 +476,38 @@ def export_freevars(self, parent, child):
 
     def reconstruct(self, codegen):
-        flags = 0x00
+        create_nested_fn_name = unique_id("__create_nested_fn")
+        codegen.tx.output.install_global(
+            create_nested_fn_name, disable(_create_nested_fn)
+        )
+
+        codegen.extend_output(
+            [
+                codegen.create_load_global(
+                    create_nested_fn_name, push_null=False, add=True
+                ),
+            ]
+        )
```
I think this can just be:
```suggestion
        cg.load_import_from(__name__, "_create_nested_fn")
```

yanboliang replied:
Yes, it works! And more concise!

yanboliang commented:
@pytorchbot merge -f "irrelevant doc failures"

jeanschmidt commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

jeanschmidt commented:
please keep the PR closed so the revert job will succeed

yanboliang commented:
@pytorchbot rebase

yanboliang commented:
@pytorchbot rebase

yanboliang committed:
[Dynamo] Fix nested function resume execution (#100426)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/99665

Let me explain the root cause using the unit test I added:
* This bug is triggered when:
  * ```wrapped``` is a nested function.
  * ```wrapped``` is in another module which is different from the main function ```fn```.
  * There is a graph break inside of ```wrapped```.
* The root cause is when resuming nested function, actually we are using the outermost function(```fn``` in my example)'s global variables, but ```wrapped``` calls ```inner_func``` which is not part of ```fn```'s globals, so we have to set correct globals when nested function resume execution.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Pull Request resolved: https://github.com/pytorch/pytorch/pull/100426

Reviewed By: angelayi

Differential Revision: D45715594

Pulled By: yanboliang

fbshipit-source-id: f0775b0b34711ce055e704c9df79168f0bafc8d4

yanboliang commented:
@pytorchbot merge
